wanda/forge.go

```diff
@@ -0,0 +1,193 @@
+package wanda
+
+import (
+	"fmt"
+	"log"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+)
+
+// Build builds a container image from the given specification file.
+func Build(specFile string, config *ForgeConfig) error {
+	if config == nil {
+		config = &ForgeConfig{}
+	}
+
+	spec, err := parseSpecFile(specFile)
+	if err != nil {
+		return fmt.Errorf("parse spec file: %w", err)
+	}
+
+	forge, err := NewForge(config)
+	if err != nil {
+		return fmt.Errorf("make forge: %w", err)
+	}
+	return forge.Build(spec)
+}
+
+// ForgeConfig is a configuration for a forge to build container images.
+type ForgeConfig struct {
+	WorkDir    string
+	DockerBin  string
+	WorkRepo   string
+	NamePrefix string
+	BuildID    string
+
+	ReadOnlyCache bool
+}
+
+// Forge is a forge to build container images.
+type Forge struct {
+	config *ForgeConfig
+
+	workDir string
+
+	remoteOpts []remote.Option
+}
+
+// NewForge creates a new forge with the given configuration.
+func NewForge(config *ForgeConfig) (*Forge, error) {
+	absWorkDir, err := filepath.Abs(config.WorkDir)
+	if err != nil {
+		return nil, fmt.Errorf("abs path for work dir: %w", err)
+	}
+
+	return &Forge{
+		config:  config,
+		workDir: absWorkDir,
+		remoteOpts: []remote.Option{
+			remote.WithAuthFromKeychain(authn.DefaultKeychain),
+		},
+	}, nil
+}
+
+func (f *Forge) addSrcFile(ts *tarStream, src string) {
+	ts.addFile(src, nil, filepath.Join(f.workDir, src))
+}
+
+func (f *Forge) workTag(name string) string {
```

Comment:
what is the tag used for?

Reply:
this is the tag / image name to be saved in the work repo, prefixed with build id.